### DISCUSSION TITLE
Finalizers behavior on Exception

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/dotnet-operator-sdk#226] [Q&A] [answered]
Hello, thank you for this great SDK.

I have a question about finalizers' behavior - what happens if the FinalizeAsync throws an exception? Is finalizer removed, or will finalization be re-scheduled and repeated again?

**Comment by erin-allison** [comment, marked as answer]
I don't have anything set up right now to properly test this, but from digging through the code, I am relatively confident that the event will be re-queued and your finalizer will be called again.

Based on the logic in `FinalizerManager<TEntity>`, we can see that if any of your registered `IResourceFinalizer<TEntity>` instances throw an exception, they will not be removed from the resource, continuing to block deletion.

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Finalizer/FinalizerManager%7BTEntity%7D.cs#L83-L91

Tracing up the call stack to `ManagedResourceController<TEntity>`, we can see `WatcherEvents` has logic specifically to handle Finalization...

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Controller/ManagedResourceController%7BTEntity%7D.cs#L78-L82

... and `ErroredEvents` mirrors that logic as well, suggesting that the event would go back on the stack to be processed again.

https://github.com/buehler/dotnet-operator-sdk/blob/02bb7ecd4309cd26ec4b3cdd60b40584fd633b94/src/KubeOps/Operator/Controller/ManagedResourceController%7BTEntity%7D.cs#L136-L142

**Reply by buehler** [reply]
Perfectly summarized :-)
Be aware, that if a finalizer constantly throws exceptions (for example wrong code) this may block the deletion of resources permanently.
